File: internships/tinkoff/25_spring/D.py
import sys


def gcd(a: int, b: int) -> int:
    while b > 0:
        a, b = b, a % b

    return a


def lcm(a: int, b: int) -> int:
    return (a // gcd(a, b)) * b


def to_add(num: int, mod: int) -> int:
    return (mod - (num % mod)) % mod


def main() -> None:
    n, x, y, z = map(int, sys.stdin.readline().split())
    arr = tuple(map(int, sys.stdin.readline().split()))

    lcm_xy = lcm(x, y)
    lcm_xz = lcm(x, z)
    lcm_yz = lcm(y, z)
    lcm_xyz = lcm(lcm_xy, z)

    dp = [[float("inf") for _ in range(8)] for _ in range(n)]
    for i in range(n):
        num = arr[i]
        dp[i][1] = to_add(num, x)
        dp[i][2] = to_add(num, y)
        dp[i][4] = to_add(num, z)
        dp[i][3] = to_add(num, lcm_xy)
        dp[i][5] = to_add(num, lcm_xz)
        dp[i][6] = to_add(num, lcm_yz)
        dp[i][7] = to_add(num, lcm_xyz)

    for i in range(1, n):
        new = [float("inf") for _ in range(8)]
        for m1 in range(1, 8):
            new[m1] = min(dp[i][m1], dp[i - 1][m1])
        for m2 in range(1, 8):
            for m3 in range(1, 8):
                m4 = m2 | m3
                new[m4] = min(new[m4], dp[i][m3] + dp[i - 1][m2])
        dp[i] = new

    print(dp[n - 1][7])


if __name__ == "__main__":
    main()

File: internships/tinkoff/25_spring/D.cpp
#include <iostream>
#include <algorithm>
#include <limits>
#include <vector>

using namespace std;

typedef long long ll;

ll gcd(ll a, ll b) {
    while (b > 0) {
        a = a % b;
        swap(a, b);
    }
    return a;
}

ll lcm(ll a, ll b) {
    return (a / gcd(a, b)) * b;
}

ll to_add(ll num, ll mod) {
    return (mod - (num % mod)) % mod;
}

int main() {
    std::ios_base::sync_with_stdio(0);
    std::cin.tie(0);
    std::cout.tie(0);

    int n;
    ll x, y, z;
    cin >> n >> x >> y >> z;
    
    vector<ll> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    
    ll lcm_xy = lcm(x, y);
    ll lcm_xz = lcm(x, z);
    ll lcm_yz = lcm(y, z);
    ll lcm_xyz = lcm(lcm_xy, z);
    
    vector<vector<ll>> minim(n, vector<ll>(8, numeric_limits<ll>::max()));
    for (int i = 0; i < n; i++) {
        ll num = arr[i];
        minim[i][1] = to_add(num, x);
        minim[i][2] = to_add(num, y);
        minim[i][4] = to_add(num, z);
        minim[i][3] = to_add(num, lcm_xy);
        minim[i][5] = to_add(num, lcm_xz);
        minim[i][6] = to_add(num, lcm_yz);
        minim[i][7] = to_add(num, lcm_xyz);
    }
    
    for (int i = 1; i < n; i++) {
        vector<ll> new_minim(8, numeric_limits<ll>::max());
        for (int m1 = 1; m1 < 8; m1++) {
            new_minim[m1] = min(minim[i][m1], minim[i - 1][m1]);
        }
        for (int m2 = 1; m2 < 8; m2++) {
            for (int m3 = 1; m3 < 8; m3++) {
                int m4 = m2 | m3;
                new_minim[m4] = min(new_minim[m4], minim[i][m3] + minim[i - 1][m2]);
            }
        }
        minim[i] = new_minim;
    }
    
    cout << minim[n - 1][7] << endl;
    return 0;
}
